genmc-sys/cpp/src/MiriInterface/Mutex.cpp

```diff
@@ -0,0 +1,142 @@
+/** This file contains functionality related to handling mutexes.  */
+
+#include "MiriInterface.hpp"
+
+// CXX.rs generated headers:
+#include "genmc-sys/src/lib.rs.h"
+
+auto MiriGenmcShim::handle_mutex_lock(ThreadId thread_id, uint64_t address, uint64_t size)
+    -> MutexLockResult {
+    // FIXME(genmc,multithreading): ensure this annotation id is unique even if Miri runs GenMC mode
+    // multithreaded. We cannot use the address directly, since it is 64 bits, and `ID` is an
+    // `unsigned int`.
+    ModuleID::ID annot_id;
+    auto [it, inserted] = annotation_id.try_emplace(address, annotation_id_counter);
+    if (inserted) {
+        annot_id = annotation_id_counter++;
+    } else {
+        annot_id = it->second;
+    }
+    const auto annot = std::move(Annotation(
+        AssumeType::Spinloop,
+        Annotation::ExprVP(NeExpr<AnnotID>::create(
+                               RegisterExpr<AnnotID>::create(ASize(size).getBits(), annot_id),
+                               ConcreteExpr<AnnotID>::create(ASize(size).getBits(), SVal(1))
+        )
+                               .release())
+    ));
+
+    // Mutex starts out unlocked, so we always say the previous value is "unlocked".
+    const auto old_val = SVal(0);
```

Comment:
This comment makes no sense at all. This is the usual CAS from 0 to 1 that acquire a mutex, isn't it? It has nothing to do with whether a mutex starts out locked or unlocked upon initialization.

Reply:
An unlocked mutex is indicated by the value being `0`, but `old_val` is _not_ used as the "expected" value for a CAS however (this is just the load part of a CAS, no comparison happens during the load).
This value is used for the temporal mixing of atomic/non-atomic accesses from the previous PR. We inform GenMC, that _if_ there is no previous atomic store to this location, `old_val` is the value that should be used (which means this is the initial value we want to give any mutex).

Reply:
Oh, that's what you mean...

That value should be entirely irrelevant since we never do normal accesses to that location?

Reply:
It is only relevant for the first `lock` or `try_lock` call, since those would otherwise read uninit memory (they need to read `SVal(0)` so they can succeed).
For `unlock`, it shouldn't matter (since it should always happen _after_ a `lock` call), so we could also pass `std::nullopt` here.

Reply:
How do these locations usually get their values? Does genmc usually require a pthread_mutex_init call? Mutexes can have static initializers though and those don't have to be all-zero...

Reply:
I think I understand now what happens; we need to just work on the comment a bit.
```suggestion
    // As usual, we need to tell GenMC which value was stored at this location before this atomic
    // access, if there previously was a non-atomic initializing access. We just always set this to 0,
    // indicating that the lock is not locked.
    const auto old_val = SVal(0);
```